src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs

```diff
@@ -0,0 +1,141 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using ILLink.RoslynAnalyzer.DataFlow;
+using ILLink.Shared.DataFlow;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using MultiValue = ILLink.Shared.DataFlow.ValueSet<ILLink.Shared.DataFlow.SingleValue>;
+using StateValue = ILLink.RoslynAnalyzer.DataFlow.LocalState<ILLink.Shared.DataFlow.ValueSet<ILLink.Shared.DataFlow.SingleValue>>;
+
+namespace ILLink.RoslynAnalyzer.TrimAnalysis
+{
+	public class TrimAnalysisVisitor : LocalDataFlowVisitor<MultiValue, ValueSetLattice<SingleValue>>
+	{
+		public readonly TrimAnalysisPatternStore TrimAnalysisPatterns;
+
+
+		public TrimAnalysisVisitor (
+			LocalStateLattice<MultiValue, ValueSetLattice<SingleValue>> lattice,
+			OperationBlockAnalysisContext context
+		) : base (lattice, context)
+		{
+			TrimAnalysisPatterns = new TrimAnalysisPatternStore ();
+		}
+
+		// Override visitor methods to create tracked values when visiting operations
+		// which reference possibly annotated source locations:
+		// - invocations (for annotated method returns)
+		// - parameters
+		// - 'this' parameter (for annotated methods)
+		// - field reference
+
+		public override MultiValue VisitInvocation (IInvocationOperation operation, StateValue state)
+		{
+			// Base logic takes care of visiting arguments, etc.
+			base.VisitInvocation (operation, state);
+
+			return new MultiValue (new SymbolValue (operation.TargetMethod, isMethodReturn: true));
```

Comment:
Performance wise: I wonder how is this going to stack up against the current linker implementation.
In the linker we basically ignore values from "unintristing" methods (or actually anything), meaning we return "Unknown" (or "Top"). The values are really only tracked when they are interesting in some way.
But here, we're attaching a symbol to everything, and fully track everything.

Reply:
Maybe we should use a factory method to create the SymbolValue and if the value is not interesting in any way, just return a singleton. That would more or less map to what linker is doing currently.

Reply:
I left a comment here, to be addressed in a later change